streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -16,171 +16,256 @@
-                final StreamThread.TaskCreator taskCreator,
-                final StreamThread.StandbyTaskCreator standbyTaskCreator,
-                final Admin adminClient,
-                final AssignedStreamsTasks active,
-                final AssignedStandbyTasks standby) {
+                final StreamThread.AbstractTaskCreator<? extends Task> taskCreator,
+                final StreamThread.AbstractTaskCreator<? extends Task> standbyTaskCreator,
+                final InternalTopologyBuilder builder,
+                final Admin adminClient) {
         this.changelogReader = changelogReader;
         this.processId = processId;
         this.logPrefix = logPrefix;
-        this.streamsMetadataState = streamsMetadataState;
-        this.restoreConsumer = restoreConsumer;
         this.taskCreator = taskCreator;
         this.standbyTaskCreator = standbyTaskCreator;
-        this.active = active;
-        this.standby = standby;
-
+        this.builder = builder;
         final LogContext logContext = new LogContext(logPrefix);
 
-        this.log = logContext.logger(getClass());
+        log = logContext.logger(getClass());
 
         this.adminClient = adminClient;
     }
 
+    void setConsumer(final Consumer<byte[], byte[]> consumer) {
+        this.consumer = consumer;
+    }
+
+    public UUID processId() {
+        return processId;
+    }
+
+    InternalTopologyBuilder builder() {
+        return builder;
+    }
+
     public Admin adminClient() {
         return adminClient;
     }
 
-    void createTasks(final Collection<TopicPartition> assignment) {
-        if (consumer == null) {
-            throw new IllegalStateException(logPrefix + "consumer has not been initialized while adding stream tasks. This should not happen.");
-        }
+    public void handleRebalanceStart() {
+        rebalanceInProgress = true;
+    }
+
+    public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                                 final Map<TaskId, Set<TopicPartition>> standbyTasks) {
+        final Map<TaskId, Set<TopicPartition>> activeTasksToCreate = new TreeMap<>(activeTasks);
 
-        if (!assignment.isEmpty() && !assignedActiveTasks.isEmpty()) {
-            resumeSuspended(assignment);
+        final Map<TaskId, Set<TopicPartition>> standbyTasksToCreate = new TreeMap<>(standbyTasks);
+
+        // first rectify all existing tasks
+        final LinkedHashMap<TaskId, RuntimeException> taskCloseExceptions = new LinkedHashMap<>();
+        final Iterator<Task> iterator = tasks.values().iterator();
+        while (iterator.hasNext()) {
+            final Task task = iterator.next();
+            if (activeTasks.containsKey(task.id()) && task.isActive()) {
+                task.resume();
+                activeTasksToCreate.remove(task.id());
+            } else if (standbyTasks.containsKey(task.id()) && !task.isActive()) {
+                task.resume();
```

Comment:
Hmm, here we know that task.resume would be a no-op so why call it? I cannot tell if it is relevant to make unit tests easier.

Reply:
I see your rationale. Maybe this code is misleading, but the idea is still that we don't worry in TaskManager about the implementation of suspend() and resume()